### ISSUE TITLE
How to train feedforward neural network using vowpal wabbit python wrapper

### ISSUE THREAD

**Issue by andrei-kondakov** [JohnLangford/vowpal_wabbit#1232]
I would like to train `the feedforward neural network `using python wrapper for vowpal wabbit tool.

```
from vowpalwabbit import pyvw
from sklearn.metrics import precision_recall_fscore_support

def train_test_vw(category_name):
    vw = pyvw.vw(binary=True,
                 quiet=False,
                 loss_function='logistic',
                 passes=4
                 nn=4,
                 c=True,
                 l=0.2)
    train_examples = get_sentences('train', category_name)
    for example_str in train_examples:
        ex = vw.example(example_str) # example_str = '-1 | football messi ronaldo'
        ex.learn()
        ex.finish()

    test_examples = get_sentences('test', category_name)
    y_pred = []
    y_true = []
    for example_str in test_examples:
        y_true.append(int(example_str.split('|')[0]))
        y_pred.append(vw.predict(example_str))
    return precision_recall_fscore_support(y_true, y_pred, average='binary')
```

But I had a bad results:
```
        category         precision  recall  f1_score
SPORT                    0.0000   0.0000   0.0000
```
Vowpal wabbit output:
> Num weight bits = 18
> learning rate = 0.2
> initial_t = 0
> power_t = 0.5
> using cache_file = .cache
> ignoring text input in favor of cache input
> num sources = 1
> /usr/local/lib/python3.6/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 due to no predicted samples.
>   'precision', 'predicted', average, warn_for)
>  
> finished run
> number of examples = 0
> weighted example sum = 0.000000
> weighted label sum = 0.000000
> average loss = nan
> total feature number = 0

I have a misunderstanding on several issues:
1. Parameter passes does not affect anything.
2. How correctly to counter the training passes and set the number of passes?
3. Why I had average loss = nan in vowpal output?
4. What exactly does ex.finish() and where it should be in my code above?
5. How to control training? To find out whether my model has not been retrained...

**Comment by JohnLangford** [comment]
Not sure what's going wrong here.  On the commandline, '--nn 4' is a just-works addition to things.   @gramhagen thoughts?

**Comment by Scott-Graham-Bose** [comment]
You cannot use the 'passes' parameter without providing a data file (using the 'd' argument) in initialization, it is trying to keep stdin open which is preventing the model from using the examples your are sending in.

If you need to do multiple passes you can loop over the training examples the desired number of times.

**Comment by Scott-Graham-Bose** [comment]
try this:
```
from vowpalwabbit import pyvw
from sklearn.metrics import precision_recall_fscore_support

def train_test_vw(category_name):
    vw = pyvw.vw(binary=True,
                 quiet=False,
                 loss_function='logistic',
                 nn=4,
                 c=True,
                 l=0.2)
    train_examples = get_sentences('train', category_name)
    for _ in range(4):
        for example_str in train_examples:
            ex = vw.example(example_str) # example_str = '-1 | football messi ronaldo'
            ex.learn()
            ex.finish()

    test_examples = get_sentences('test', category_name)
    y_pred = []
    y_true = []
    for example_str in test_examples:
        y_true.append(int(example_str.split('|')[0]))
        y_pred.append(vw.predict(example_str))
    return precision_recall_fscore_support(y_true, y_pred, average='binary')

**Comment by Scott-Graham-Bose** [comment]
also, i'm pretty sure you can use the model to build the examples
```
 ex = vw.example(example_str) # example_str = '-1 | football messi ronaldo'
 ex.learn()
 ex.finish()
```
would just be
```
vw.learn(example_str)
```

**Comment by Scott-Graham-Bose** [comment]
I can also add 'nn'  into the slkearn model, to make it easier to train / evaluate

**Comment by JohnLangford** [comment]
It may be worth adding 'nn' if that's easy.  I've certainly found it a useful option a few times.

**Comment by JohnLangford** [comment]
Patch from @gramhagen merged in, so closing.
